src/renderer/Dataframe.js

```diff
@@ -135,27 +105,47 @@ export default class Dataframe {
     }
 
     inViewport(featureIndex, scale, center, aspect) {
-        const { minx, miny, maxx, maxy } = this._getBounds(scale, center, aspect);
+        return this._geometryInViewport(featureIndex, scale, center, aspect); // FIXME
+    }
+
+    _geometryInViewport(featureIndex, scale, center, aspect) {
+        const height = scale * (center.y - this.center.y);
+        const width = (scale / aspect) * (center.x - this.center.x);
+        const viewport = this._getViewportGeometry(center, width, height);
+        const triangles = this._featureGeometries[featureIndex];
 
         switch (this.type) {
-            case 'point': {
-                const x = this.geom[2 * featureIndex + 0];
-                const y = this.geom[2 * featureIndex + 1];
-                return x > minx && x < maxx && y > miny && y < maxy;
-            }
+            case 'point':
+                return this._isPointInViewport(featureIndex, scale, center, aspect);
             case 'line':
-            case 'polygon': {
-                const aabb = this._aabb[featureIndex];
-                return !(minx > aabb.maxx || maxx < aabb.minx || miny > aabb.maxy || maxy < aabb.miny);
-            }
+            case 'polygon':
+                return this._isPolygonInViewport(triangles, viewport);
             default:
                 return false;
         }
     }
 
+    _isPointInViewport(featureIndex, scale, center, aspect) {
+        const { minx, maxx, miny, maxy } = this._getBounds(scale, center, aspect);
+        const x = this.geom[2 * featureIndex + 0];
+        const y = this.geom[2 * featureIndex + 1];
+        return x > minx && x < maxx && y > miny && y < maxy;
+    }
+
+    _isPolygonInViewport(triangles, viewport) {
+        for (let i = 0; i < triangles.length; i++) {
+            if (triangles[i].collides(viewport)) {
```

Comment:
:confused: The idea of using an AABB is to actually avoid this collision check most of the times, like this:
```
_isPolygonInViewport(triangles, viewport) {
        const result = checkAABB(polygonAABB, viewport);
        if (result !==undefined){
               return result;
       } 
        for (let i = 0; i < triangles.length; i++) {
               ....
        }
}
```

Reply:
Yes, you are right!